feat(functional-tests): Functional tests for reset password with code

vpomerleau opened the pull request:
## Because

* We have a new password reset flow that uses confirmation codes instead of link
## This pull request

* Remove backbone resetPassword playwright page object model (POM), no longer in use
* Move all reset password tests out of react-conversion folder, and into new resetPassword folder with subfolders for resetPasswordWithLink and resetPasswordWithCode
* Mark the functional tests for reset password with link for deletion with ticket number
* Add new tests updated for reset password with code flow in resetPasswordWithCode folder
* Fix bug with password strength bubble not displayed due to discontinued onFocus metric

## Issue that this pull request solves

Closes: #FXA-9612

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

pdehaan commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/oauthResetPassword.spec.ts:
```diff
@@ -0,0 +1,197 @@
+      // TODO in FXA-8006 - Update this test to use the 123Done PKCE button once fixed
+      // For now navigate directly to the link.
+      await page.goto(
+        `http://localhost:3030/authorization?` +
```
Should this be `localhost:3030` or should it be env-specific? (ie: different URLs for stage/prod)

vpomerleau replied:
Good callout. I pulled this test from the reset with link version, but it was previously skipped so maybe had never run against stage/prod. Will update with `target.contentServerUrl`

pdehaan commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/oauthResetPassword.spec.ts:
```diff
@@ -0,0 +1,197 @@
+      // For now navigate directly to the link.
+      await page.goto(
+        `http://localhost:3030/authorization?` +
+          `&access_type=offline` +
```
tiny nit: this will generate a search query like `/authorization?&access_type=` (with `?&`).
Possibly not worth creating a new `URL()` instance and setting `query` param if an object syntax is easier to read.

```js
const uri = new URL("/authorization", "http://localhost:3030");
uri.search = new URLSearchParams({
  beep: 'boop',
  foo: 'bar',
});
page.goto(uri.toString());
```

vpomerleau replied:
This is meant to be temporary since we have a ticket filed to fix the 123Done button - I'll just remove the stray `&` on the first param for now.

pdehaan commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/resetPassword.spec.ts:
```diff
@@ -0,0 +1,166 @@
+      config.featureFlags.resetPasswordWithCode !== true,
+      'see FXA-9728, remove conditional skip when feature flag removed'
+    );
+    test.slow();
```
nit, does this need the `test.slow()` now that @Trinaa tweaked the global test timeout?

Trinaa replied:
Yep. Make sure to remove this!

pdehaan commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/resetPassword.spec.ts:
```diff
@@ -0,0 +1,166 @@
+
+    resetPasswordReact.resendButton.click();
+
+    await expect(resetPasswordReact.statusBar).toHaveText(/Email re-?sent/);
```
😍

Trinaa reviewed (changes requested):
I like the organization and user-facing locator improvements! 🎉 

Blocking because I see a lot of comments with 'TODO in FXA-9612'. Just want to make sure these things aren't meant to be done as part of this ticket?
<img width="741" alt="image" src="https://github.com/mozilla/fxa/assets/3422688/056108cc-acda-4e0a-99d7-81d35d542493">

I think it may be worth it to rebase on main on this PR to make sure none of the `test.slow` annotations get through.

vpomerleau commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/oauthResetPassword.spec.ts:
```diff
@@ -0,0 +1,197 @@
+        resetPasswordReact.passwordResetConfirmationHeading
+      ).toBeVisible();
+
+      // TODO in FXA-9561 - Verify that the service name is displayed in the "Continue to ${serviceName}" button
+      // This functionality is not yet implemented in the reset password flow
+
+      // TODO in FXA-9561 - Remove this temporary test of sign in with new password
+      // user should be signed in and able to navigate to the relying party
```
The test correctly reflects the current and expected state of the flow, so I don't think it should be skipped. The notes here were reminders for test changes that will be needed when we work on the next scope of work for password reset redesign.  I can leave it out if not best practice.

Trinaa replied:
Sounds like it is well considered, so I'm good to leave things as they are!

vpomerleau commented on packages/functional-tests/pages/resetPasswordReact.ts:
```diff
@@ -7,9 +7,9 @@ export class ResetPasswordReactPage extends BaseLayout {
-        .getByRole('heading', { name: /^Reset password/ })
-        // for password reset redesign, with resetPasswordWithCode flag
-        .or(this.page.getByRole('heading', { name: /^Password reset/ }))
+        .getByRole('heading', { name: /^Reset your password/ })
+        // TODO in FXA-9728, remove the .or (support for reset password with link)
+        .or(this.page.getByRole('heading', { name: /^Reset password/ }))
```
Thanks for the suggestion! I'll update all the locators that have this format in a grouped PR update commit (+ the heading text changed so the the recommendation above doesn't have latest content 🙃 )

vpomerleau commented on packages/functional-tests/pages/resetPasswordReact.ts:
```diff
@@ -127,10 +145,12 @@ export class ResetPasswordReactPage extends BaseLayout {
-    await this.beginResetButton.click();
+  async fillOutResetPasswordCodeForm(code: string) {
+    await expect(this.confirmResetPasswordHeading).toBeVisible();
+    // using .pressSequentially() instead of .fill()
+    // because of onChange handling on the code input component
+    // that distributes typed text or pasted text between single digit inputs
+    await this.confirmationCodeFirstInput.pressSequentially(code);
```
In this instance, I think this is a good use case for `pressSequentially` since there is custom keyboard handling for these inputs.

vpomerleau commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/resetPassword.spec.ts:
```diff
@@ -0,0 +1,166 @@
+
+import { expect, test } from '../../../lib/fixtures/standard';
+
+// This test files contains the latest tests for reset password with code (React)
+// See FXA-9344
```
This comment can be removed entirely, I think.

vpomerleau commented on packages/functional-tests/tests/resetPassword/resetPasswordWithCode/resetPassword.spec.ts:
```diff
@@ -0,0 +1,166 @@
+  test('visit confirmation screen without initiating reset_password, user is redirected to /reset_password', async ({
+    pages: { resetPasswordReact },
+  }) => {
+    test.fixme(true, 'TODO in FXA-9612, should be fixed when FXA-9722 merged');
```
Yes! Test now passes.

vpomerleau commented:
> I like the organization and user-facing locator improvements! 🎉

🙌 

> Blocking because I see a lot of comments with 'TODO in [FXA-9612](https://mozilla-hub.atlassian.net/browse/FXA-9612)'. Just want to make sure these things aren't meant to be done as part of this ticket?

Thanks for calling this out, I'd missed the key-stretching test updates. Added now, and updated other comments with the correct ticket numbers where appropriate (or removed comments entirely)

Trinaa reviewed (approved):
LGTM! 🚢
